### PR TITLE
Update scala notes

### PR DESCRIPTION
**Pull request by olivierschultz**
* ADT
* containers
* evaluation

### CODE REVIEW

File: scala/adt.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+Algebraic Data Types
+====================
+
+ADT are just a way of modelling data in terms of two patterns:
+* logical ors, such as `List`, whish is `:` or `Nil` (sum type)
```
Comment (BerthelotAnthony) [inline scala/adt.md]: which

File: scala/adt.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+## Structural Recursion
+
+```
+anOption match {
+  case None    => doNoneCase
+  case Some(x) => doSomeCase(x)
+}
+```
+
+We can generalise this pattern to any ADT. Each branch is a sum type (a logical or) gets its own `¢ase` in the pattern matching, an each product type (a logical and) requires us to extract and do
```
Comment (BerthelotAnthony) [inline scala/adt.md]: generalize ? I believe it's written with a Z but not sure you know yeah you know dab

File: scala/adt.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+## Exhaustiveness Checking
+
+When we define an ADT using `sealed` traits we allow the compiler to perform ehaustiveness checking. In simpler words, this means the compiler will shout at us if we miss out a case in our structural recursion.
+
+## The Inner Life of Sealed
+
+A `sealed` trait can only be extended within the file in which it defined. This allows the compiler to perform exhaustiveness checking for pattern matches on the trait. Why? Because the compiler must know all the possible subtypes to do the checks, and given that the JVM allows code to be loaded at runtime the compiler can't scan the whole program to collect all the possible subtypes.
```
Comment (BerthelotAnthony) [inline scala/adt.md]: in which it IS defined ???

File: scala/adt.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  case None    => doNoneCase
+  case Some(x) => doSomeCase(x)
+}
+```
+
+We can generalise this pattern to any ADT. Each branch is a sum type (a logical or) gets its own `¢ase` in the pattern matching, an each product type (a logical and) requires us to extract and do
+
+## Exhaustiveness Checking
+
+When we define an ADT using `sealed` traits we allow the compiler to perform ehaustiveness checking. In simpler words, this means the compiler will shout at us if we miss out a case in our structural recursion.
```
Comment (BerthelotAnthony) [inline scala/adt.md]: exhau....

File: scala/adt.md (hunk 2 above)
Comment (BerthelotAnthony) [inline scala/adt.md]: and each product type...

Comment (BerthelotAnthony) [inline scala/adt.md]: > Each branch is a sum type (a logical or) gets its own `¢ase` in the

Missing word ? is a sum type gets its ... Add a coma or and ?

File: scala/containers.md [hunk 5]
```diff
@@ -20,7 +20,47 @@ final case class Some[A](a: A) extends Option[A]
...
+  final case class Only(retailer: Retailer) extends RetailerCriteria
+}
+
+def searchOffers(
+  product: SearchCriteria,
+  retailer: RetailerCriteria
+  ): Seq[Offer] = ???
+```
+
+You can also easily extends your ADT to new functionality.
```
Comment (BerthelotAnthony) [inline scala/containers.md]: you can extend without s

File: scala/containers.md [hunk 6]
```diff
@@ -20,7 +20,47 @@ final case class Some[A](a: A) extends Option[A]
...
+_Note_: 
+
+* Option can be pattern matched
+
+
+### Don't overload the semantic
+
+Don't overload the semantic of `None` and `Some`. The former is very similar to how `null` in java is sometimes used with meanings that are different from simple absence. 
+
+Whenever someone in your team is confused about the exact meaning of an `Option` in your code, this is a good indicator that you should introduce your own ADT that captures the implicit semanics of your domain.
```
Comment (BerthelotAnthony) [inline scala/containers.md]: semantics

File: scala/evaluation.md [hunk 7]
```diff
@@ -1,6 +1,11 @@
 Evaluation
 ==========
 
+A non-primitive expresion is evaluated as follows.
```
Comment (BerthelotAnthony) [inline scala/evaluation.md]: expression

File: scala/containers.md (hunk 5 above)
Comment (olivierschultz) [reply scala/containers.md]: It refers to the typology of the keyword in Scala which is `extends`

File: scala/evaluation.md [hunk 8]
```diff
@@ -24,3 +29,24 @@ _Note_: One good reason for choosing `def` over `val`, especially in abstract cl
...
+3 * 3 + square(4)
+9 + square(4)
+9 + 4 * 4
+9 + 16
+25
+```
+
+This scheme of expression is called _substitution model_.
+
+The idea underlying this model is that all evaluation does is reduce an expression to a value.
```
Comment (BerthelotAnthony) [inline scala/evaluation.md]: all evualation doesi is reducING ??? Not sure I never know this type of stuff  so .... Don't take this remark for granted ty

File: scala/adt.md (hunk 2 above)
Comment (olivierschultz) [reply scala/adt.md]: Both are correct. But I'll go with generalize since it's more common to non English-speaking

File: scala/syntax.md [hunk 9]
```diff
@@ -26,6 +26,15 @@ for (i <= 0 to 10) {
 
 _Note_: Guards are simply boolea expressions which are used to make cases more specific.
```
Comment (BerthelotAnthony) [inline scala/syntax.md]: boolean

**Pull request merged**